File: assistant.py
from typing import Iterable
from openai.types.chat import ChatCompletion
from openai.types.chat.chat_completion_user_message_param import ChatCompletionUserMessageParam
from openai import OpenAI, Stream
from openai.types.chat.chat_completion_message_param import ChatCompletionMessageParam
import env
from pprint import pprint
import httpx
import json

import http.client

class Assistant:
    def __init__(self, temperature = 0.1, model = "meta-llama/Meta-Llama-3-8B-Instruct", system_prompt = "", local = False, messages = []) -> None:
        self.temperature = temperature
        self.model = model
        self.description = system_prompt
        self.top_p = 1
        if local:
            self.api_key = "http://localhost:11434/v1"
            self.base_url = "ollama"
        else:
            self.api_key = env.api_key
            self.base_url = env.base_url

        self.messages = messages

    def run_stream(self, question: str) -> Iterable[str]:
        self.messages.append({
            "role": "user",
            "content": question
        })

        messages = [
            {
                "role": "system",
                "content": self.description
            },
            *self.messages
        ]

        res = httpx.post(
            self.base_url + "/chat/completions",
            headers={"Authorization": "bearer " + self.api_key, "Content-Type": "application/json"},
            json={
                "model": self.model,
                "messages": messages,
                "temperature": self.temperature,
                "top_p": self.top_p,
                "stream": True,
                "max_tokens": 8192
            }
        )

        response = ""
        for chunk in res:
            if chunk.choices[0].delta.content is not None:
                response += chunk.choices[0].delta.content
                yield chunk.choices[0].delta.content

        self.messages.append({
            "role": "assistant",
            "content": response
        })

    def run(self, question: str) -> str:
        self.messages.append({
            "role": "user",
            "content": question
        })

        messages = [
            {
                "role": "system",
                "content": self.description
            },
            *self.messages
        ]

        res = httpx.post(
            self.base_url + "/chat/completions",
            headers={"Authorization": "bearer " + self.api_key, "Content-Type": "application/json"},
            json={
                "model": self.model,
                "messages": messages,
                "temperature": self.temperature,
                "top_p": self.top_p,
                "max_tokens": None
            },
            timeout = 60
        )

        response = res.json()["choices"][0]["message"]["content"]

        self.messages.append({
            "role": "assistant",
            "content": response
        })

        return response or ""

    def get_messages(self):
        return self.messages

    def clear_messages(self):
        self.messages = []

File: lge_rs/src/assistant.rs
use serde::{Serialize, Deserialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
struct Message {
    role: String,
    content: String,
}

pub struct Assistant {
    temperature: f64,
    model: String,
    description: String,
    top_p: f64,
    api_key: String,
    base_url: String,
    messages: Vec<Message>,
}

impl Assistant {
    pub fn new(model: &str, system_prompt: &str, temperature: f64) -> Assistant {
        Assistant {
            temperature,
            model: model.to_string(),
            description: system_prompt.to_string(),
            top_p: 1.,
            api_key: std::env::var("api_key").expect("API key not found"),
            base_url: std::env::var("base_url").expect("Base URL not found"),
            messages: vec![],
        }
    }
    pub fn run(&mut self, question: &str) -> Result<String, reqwest::Error> {
        self.messages.push(Message {
            role: "user".to_string(),
            content: question.to_string(),
        });

        let mut messages = vec![
            Message {
                role: "system".to_string(),
                content: self.description.to_string(),
            },
        ];
        for message in self.messages.iter() {
            messages.push(message.clone());
        }

        let client = reqwest::blocking::Client::new();

        let data = serde_json::json!({
            "model": self.model,
            "temperature": self.temperature,
            "top_p": self.top_p,
            "max_tokens": 8192,
            "messages": messages,
        });

        // println!("{:?}", messages);
        let res = client
               .post(self.base_url.to_string() + "/chat/completions")
               .header("Authorization", "Bearer ".to_string() + &self.api_key)
               .header("Content-Type", "application/json")
               .json(&data)
               .send()?;


        let body: serde_json::Value = res.json()?;
        // println!("{:?}", body);

        let answer = body["choices"][0]["message"]["content"].as_str().unwrap();

        self.messages.push(Message {
            role: "assistant".to_string(),
            content: answer.to_string(),
        });

        Ok(answer.to_string())
    }
}
